ts/spi/provider_frame.ts

```diff
@@ -544,4 +571,10 @@ export class ProviderFrame {
 
     return score;
   }
+
+  private race<T>(value: T, producer: Promise<T>): Promise<T> {
```

Comment:
I would call this method `cancellable` or similar, and rename value to cancelValue, probably move it to the second argument position too.